dockerunit-core/src/main/java/com/github/qzagarese/dockerunit/DockerUnitRule.java

```diff
@@ -0,0 +1,176 @@
+import com.github.qzagarese.dockerunit.internal.reflect.UsageDescriptorBuilder;
+import com.github.qzagarese.dockerunit.internal.service.DefaultServiceContext;
+
+public class DockerUnitRule implements TestRule {
+
+    private static final Logger logger = Logger.getLogger(DockerUnitRule.class.getSimpleName());
+
+    private static final Map<String, ServiceContext> activeContexts = new HashMap<>();
+
+    private final UsageDescriptorBuilder descriptorBuilder = DependencyDescriptorBuilderFactory.create();
+    private final ServiceContextBuilder contextBuilder = ServiceContextBuilderFactory.create();
+    private final DiscoveryProvider discoveryProvider;
+    private final String serviceContextName;
+    private ServiceContext discoveryContext;
+    
+    private static final String GLOBAL_CONTEXT_NAME = "dockerunit_global_context";    
+
+    /**
+     * Returns the {@link ServiceContext} that has been associated to the instance of {@link DockerUnitRule} that
+     * has serviceContextName as parameter.
+     * 
+     * @param serviceContextName
+     * @return the associated {@link ServiceContext}
+     */
+    public static ServiceContext getServiceContext(String serviceContextName) {
+        return Optional.ofNullable(activeContexts.get(serviceContextName))
+                .orElseThrow(() -> new ConfigException("No active context for name " + serviceContextName 
+                        + " . Please make sure that you are using the context name that you have passed to the "
+                        + DockerUnitRule.class.getSimpleName() 
+                        + " constructor."));
+    }
+    
+    /**
+     * Returns the default {@link ServiceContext} or a random one if all the current instances of {@link DockerUnitRule}
+     * have been initialised with a specific service context name.
+     * 
+     * @return the default{@link ServiceContext}
+     */
+    public static ServiceContext getDefaultServiceContext() {
+        return Optional.ofNullable(activeContexts.get(GLOBAL_CONTEXT_NAME))
+                .orElse(
+                        activeContexts.values().stream().findAny()
+                                .orElseThrow(() -> new ConfigException("No active context detected. "
+                                    + "Please make sure that you have declared at least one @"
+                                    + Use.class.getSimpleName()
+                                    + " annotation on tha class that declares your "
+                                    + DockerUnitRule.class.getSimpleName()
+                                    + " instance.")));
+    }
+    
+    public DockerUnitRule() {
+        this(GLOBAL_CONTEXT_NAME);
+    }
+    
+    public DockerUnitRule(String serviceContextName) {
+        ServiceLoader<DiscoveryProviderFactory> loader = ServiceLoader.load(DiscoveryProviderFactory.class);
+       
+        this.discoveryProvider = StreamSupport.stream(loader.spliterator(), false)
+                .map(impl -> {
```

Comment:
you can use `peek` here if you want since the purpose of this instruction is just log something.

Reply:
thanks!